### PR TITLE
Add ability to use paratest.chapcs from nightly, use that for valgrind

### PR DESCRIPTION
**Pull request by e-kayrakli**
The main purpose of this PR is to use paratest.chapcs for valgrind testing so
that it runs quicker. In order to do that this PR:

- Adds a `-chapcs` flag to `nightly`

- Parametrizes the script that `nightly` uses so that we can choose between
  `start_test`, `paratest.server`, `paratest.chapcs` more easily

- Adds a `-nightly` flag to `paratest.chapcs`

- Adjusts `paratest.chapcs` to detect `-chapcs` argument anywhere in the command
  line and if it exists:

  - To stop checking for the exclusive nightly hours
  - To stop adding `CHPL_TEST_LIMIT_RUNNING_EXECUTABLES` (this'll require us to
    use exclusive allocations for correctness tests on chapcs)

- While there, adds a global `debug = False` that can be switched on to run the
  script locally without slurm to generate some debugging output.

- Drops localnode stuff from valgrind cron script, and throws in `-chapcs` while
  calling `nightly`

I have considered using `argparse` in `paratest.chapcs` but it may be an overkill.
Also, its prefix matching can only be disabled after Python 3.5 and I didn't want to
think whether that's an issue.

### CODE REVIEW

File: util/test/paratest.chapcs [hunk 1]
```diff
@@ -133,17 +134,27 @@ def get_good_nodepara():
...
     Run paratest inside an salloc using all nodes that are not reserved
     exclusively on the chapel partition. Throw `--oversubscribe --nice` and
     turn off affinity and limit how many executables can run at once so we play
     nice with other testing going on.
     """
-    nodepara = get_good_nodepara()
-    num_free_nodes = get_num_non_exclusive_nodes()
-    para_env = ['-env', 'CHPL_TEST_LIMIT_RUNNING_EXECUTABLES=yes', '-env',
-                'CHPL_RT_OVERSUBSCRIBED=yes']
+    if not debug:
```
Comment (ronawho) [inline util/test/paratest.chapcs:144]: Can we rename `debug` to `dry_run` or `debug_no_slurm` or something and pull it into a different PR? Personally, I  think the debug branches make the script harder to follow and most of the script is slurm commands so I don't mind debugging/running on chapcs, but if it was useful for debugging I don't mind keeping it.

File: util/test/paratest.chapcs [hunk 2]
```diff
@@ -133,17 +134,27 @@ def get_good_nodepara():
...
+    else:
+        print('Running in debug mode')
+        print('paratest_args: {}'.format(paratest_args))
+        print('nightly: {}'.format(nightly))
+        nodepara = 1
+        num_free_nodes = 1
+
+    para_env = ['-env', 'CHPL_RT_OVERSUBSCRIBED=yes']
+    # we run exclusively with nightly, so no need to limit
+    if not nightly:
```
Comment (ronawho) [inline util/test/paratest.chapcs:156]: I think the running executable limiting is more about valgrind running than nightly being used. If we used paratest.chapcs for say nightly gasnet testing, I think we'd still want to limit the running executables. I'd probably base this off `-valgrind[exe]` being in the args -- I think that also makes sense for users doing a paratest.chapcs.

File: util/test/paratest.chapcs [hunk 3]
```diff
@@ -133,17 +134,27 @@ def get_good_nodepara():
...
+        num_free_nodes = get_num_non_exclusive_nodes()
+    else:
+        print('Running in debug mode')
+        print('paratest_args: {}'.format(paratest_args))
+        print('nightly: {}'.format(nightly))
+        nodepara = 1
+        num_free_nodes = 1
+
+    para_env = ['-env', 'CHPL_RT_OVERSUBSCRIBED=yes']
+    # we run exclusively with nightly, so no need to limit
```
Comment (ronawho) [inline util/test/paratest.chapcs:155]: I don't think we're running exclusively yet. I think this script is still throwing `--oversubscribe` (which is the new-ish replacement for the old `--share`). I think we should add an `-exclusive` flag to this script and have `nightly` throw that similar to what you did with the `-nightly` flag. Or we could have the `-nightly` flag trigger it to reinforce that exclusive access isn't really a user-facing option.

File: util/cron/nightly [hunk 4]
```diff
@@ -152,6 +158,8 @@ while (@ARGV) {
         $asserts = 1;
     } elsif ($flag eq "-mason") {
         $mason_build = 1;
+    } elsif ($flag eq "-chapcs") {
```
Comment (ronawho) [inline util/cron/nightly:161]: I'd probably call this `paratest.chapcs` and I'm thinking we'll want to be able to specify the `-nodepara` so that we can still increase the amount of oversubscription we're using for valgrind.

**Comment by ronawho** [comment]
> I have considered using argparse in paratest.chapcs but it may be an overkill.
Also, its prefix matching can only be disabled after Python 3.5 and I didn't want to
think whether that's an issue.

Hmm, I do love argparse, but the prefix matching has caused us pain in the past :( -- c599d21fcbb21ccfa837cce3157c523e7dadbc77

File: util/test/paratest.chapcs (hunk 1 above)
Comment (e-kayrakli) [reply util/test/paratest.chapcs:144]: Yeah I see your point. I like testing things as much as possible locally, though I don't know whether I think strongly enough to create another PR for that.

I'll remove this stuff

File: util/test/paratest.chapcs (hunk 3 above)
Comment (e-kayrakli) [reply util/test/paratest.chapcs:155]: The counterpart of this PR on our ci-config repo does make chapcs testing use `--exclusive`. So `-nightly` for this script would imply exclusive access. Do you still want to make it more explicit?

File: util/test/paratest.chapcs (hunk 2 above)
Comment (e-kayrakli) [reply util/test/paratest.chapcs:156]: Similar to my comment above, as we'll also switch to exclusive allocations for nightly runs, I thought we wouldn't want to limit as there won't be any other job running on the same node to mess with.

But I can still make that change if you want?

File: util/test/paratest.chapcs (hunk 3 above)
Comment (ronawho) [reply util/test/paratest.chapcs:155]: I think our 2 options were to do the slurm allocation in chapel-ci-config, and just use paratest.server here or have paratest.chpacs do the slurm reservation. I don't think we can do the salloc in both places. In chapel-ci-config, I thought we changed the correctness jobs that use a single node to run in exclusive mode.

File: util/cron/nightly (hunk 4 above)
Comment (e-kayrakli) [reply util/cron/nightly:161]: Just so I'm clear: You are suggesting that we add `-nodepara` to both `nightly` and `paratest.chapcs` and make that override `get_good_nodepara` in `paratest.chapcs`, right?

File: util/test/paratest.chapcs (hunk 2 above)
Comment (ronawho) [reply util/test/paratest.chapcs:156]: Even for a single job, we usually want to limit to only running a single executable at a time so that multiple instances of the runtime don't interfere with each other (i.e. if you run 2 `./hello`s there can be significant interference from our runtime, though we limit that with `CHPL_RT_OVERSUBSCRIBED=yes` above)  I think valgrind is the exception to that because valgrind serializes so it can't really cause interference with other executions.

File: util/cron/nightly (hunk 4 above)
Comment (ronawho) [reply util/cron/nightly:161]: Yeah, I would have nightly take a `-paratest.chapcs <nodepara>` flag and pass `-nodepara <nodepara>` to paratest.chapcs (If you just specify `-nodepara <nodepara>` to paratest.chapcs it will override whatever value paratest.chapcs calculates.

**Comment by e-kayrakli** [comment]
After a brief discussion with the rest of the team, we decided to not go this route and split valgrind into two separate jobs. See https://github.com/chapel-lang/chapel/pull/16356
